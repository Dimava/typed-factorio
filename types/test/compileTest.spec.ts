import * as path from "path"
import * as ts from "typescript"
import * as fs from "fs"
import { Transpiler } from "typescript-to-lua"

const compileTestDir = path.resolve(__dirname, "./compileTest")

function testCompiles(factorioVersion: string, fileName: string): void {
  const file = path.resolve(compileTestDir, fileName)
  const declarations = path.resolve(__dirname, `../${factorioVersion}.d.ts`)
  const rootNames = [file, declarations]
  const options = {
    target: ts.ScriptTarget.ESNext,
    moduleResolution: ts.ModuleResolutionKind.NodeJs,
    types: [],
    lib: ["lib.esnext.d.ts"],
    strict: true,
  }
  const program = ts.createProgram(rootNames, options)

  const outFiles: Array<{ fileName: string; fileContent: string }> = []
  const { diagnostics: transpileDiagnostics } = new Transpiler().emit({
    program,
    writeFile: (fileName, fileContent) => outFiles.push({ fileName, fileContent }),
  })

  const diagnostics = ts.sortAndDeduplicateDiagnostics([...ts.getPreEmitDiagnostics(program), ...transpileDiagnostics])
  expect(diagnostics.map((d) => d.messageText)).toEqual([])

  const luaFileName = fileName.substring(0, fileName.lastIndexOf(".ts")) + ".lua"
  const outFile = outFiles.find((f) => f.fileName.endsWith(luaFileName))
  expect(outFile).toBeDefined()
  expect(outFile!.fileContent.trim()).toMatchSnapshot()
}

const generatorSrc = path.resolve(__dirname, "../generatorSrc")
const versions = fs
  .readdirSync(generatorSrc)
  .map((file) => file.match(/runtime-api-([0-9-.]+)\.json/)?.[1])
  .filter((file) => !!file) as string[]

const files = fs.readdirSync(compileTestDir).filter((f) => f.endsWith(".ts"))

files.forEach((file) => {
  describe(file, () => {
    versions.forEach((v) => {
      test(`factorio version ${v}`, () => {
        testCompiles(v, file)
      })
    })
  })
})
