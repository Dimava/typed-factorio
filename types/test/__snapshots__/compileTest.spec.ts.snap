// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`events.ts factorio version 1.1.35 1`] = `
"--[[ Generated with https://github.com/TypeScriptToLua/TypeScriptToLua ]]
script.on_event(
    defines.events.on_ai_command_completed,
    function(e)
        print(e.result)
    end
)
script.on_event(defines.events.on_area_cloned, nil)
script.on_event(
    defines.events.on_built_entity,
    function(e)
        log(
            serpent.line(e)
        )
    end,
    {{filter = \\"type\\", type = \\"foo\\"}, {filter = \\"ghost_type\\", type = \\"bar\\"}}
)
script.set_event_filter(defines.events.on_robot_built_entity, {{filter = \\"circuit-network-connectable\\"}})
filters = script.get_event_filter(defines.events.on_entity_damaged)
script.on_event(
    \\"someInput\\",
    function(e)
        log(
            serpent.block(e)
        )
    end
)
myCustomEvent = script.generate_event_name()
script.on_event(
    myCustomEvent,
    function(d)
        log(d.myCustomData)
    end
)
script.get_event_handler(myCustomEvent)({myCustomData = 0})"
`;

exports[`events.ts factorio version 1.1.36 1`] = `
"--[[ Generated with https://github.com/TypeScriptToLua/TypeScriptToLua ]]
script.on_event(
    defines.events.on_ai_command_completed,
    function(e)
        print(e.result)
    end
)
script.on_event(defines.events.on_area_cloned, nil)
script.on_event(
    defines.events.on_built_entity,
    function(e)
        log(
            serpent.line(e)
        )
    end,
    {{filter = \\"type\\", type = \\"foo\\"}, {filter = \\"ghost_type\\", type = \\"bar\\"}}
)
script.set_event_filter(defines.events.on_robot_built_entity, {{filter = \\"circuit-network-connectable\\"}})
filters = script.get_event_filter(defines.events.on_entity_damaged)
script.on_event(
    \\"someInput\\",
    function(e)
        log(
            serpent.block(e)
        )
    end
)
myCustomEvent = script.generate_event_name()
script.on_event(
    myCustomEvent,
    function(d)
        log(d.myCustomData)
    end
)
script.get_event_handler(myCustomEvent)({myCustomData = 0})"
`;

exports[`events.ts factorio version 1.1.37 1`] = `
"--[[ Generated with https://github.com/TypeScriptToLua/TypeScriptToLua ]]
script.on_event(
    defines.events.on_ai_command_completed,
    function(e)
        print(e.result)
    end
)
script.on_event(defines.events.on_area_cloned, nil)
script.on_event(
    defines.events.on_built_entity,
    function(e)
        log(
            serpent.line(e)
        )
    end,
    {{filter = \\"type\\", type = \\"foo\\"}, {filter = \\"ghost_type\\", type = \\"bar\\"}}
)
script.set_event_filter(defines.events.on_robot_built_entity, {{filter = \\"circuit-network-connectable\\"}})
filters = script.get_event_filter(defines.events.on_entity_damaged)
script.on_event(
    \\"someInput\\",
    function(e)
        log(
            serpent.block(e)
        )
    end
)
myCustomEvent = script.generate_event_name()
script.on_event(
    myCustomEvent,
    function(d)
        log(d.myCustomData)
    end
)
script.get_event_handler(myCustomEvent)({myCustomData = 0})"
`;

exports[`guiElement.ts factorio version 1.1.35 1`] = `
"--[[ Generated with https://github.com/TypeScriptToLua/TypeScriptToLua ]]
button = game.player.gui.screen.add({type = \\"button\\", enabled = true, mouse_button_filter = {\\"left\\"}})
name = button.name
child = button.namee
chooseElemButton = game.player.gui.screen.add({type = \\"choose-elem-button\\", elem_type = \\"item\\", item = \\"foo\\", filters = {{filter = \\"name\\", name = \\"foo2\\"}}})"
`;

exports[`guiElement.ts factorio version 1.1.36 1`] = `
"--[[ Generated with https://github.com/TypeScriptToLua/TypeScriptToLua ]]
button = game.player.gui.screen.add({type = \\"button\\", enabled = true, mouse_button_filter = {\\"left\\"}})
name = button.name
child = button.namee
chooseElemButton = game.player.gui.screen.add({type = \\"choose-elem-button\\", elem_type = \\"item\\", item = \\"foo\\", filters = {{filter = \\"name\\", name = \\"foo2\\"}}})"
`;

exports[`guiElement.ts factorio version 1.1.37 1`] = `
"--[[ Generated with https://github.com/TypeScriptToLua/TypeScriptToLua ]]
button = game.player.gui.screen.add({type = \\"button\\", enabled = true, mouse_button_filter = {\\"left\\"}})
name = button.name
child = button.namee
chooseElemButton = game.player.gui.screen.add({type = \\"choose-elem-button\\", elem_type = \\"item\\", item = \\"foo\\", filters = {{filter = \\"name\\", name = \\"foo2\\"}}})"
`;

exports[`helloWorld.ts factorio version 1.1.35 1`] = `
"--[[ Generated with https://github.com/TypeScriptToLua/TypeScriptToLua ]]
function greetPlayer(self, player)
    player.print(\\"Hello, world!\\")
end
script.on_event(
    defines.events.on_player_created,
    function(e)
        greetPlayer(
            nil,
            game.get_player(e.player_index)
        )
    end
)
script.on_init(
    function()
        for ____, luaPlayer in pairs(game.players) do
            greetPlayer(nil, luaPlayer)
        end
    end
)"
`;

exports[`helloWorld.ts factorio version 1.1.36 1`] = `
"--[[ Generated with https://github.com/TypeScriptToLua/TypeScriptToLua ]]
function greetPlayer(self, player)
    player.print(\\"Hello, world!\\")
end
script.on_event(
    defines.events.on_player_created,
    function(e)
        greetPlayer(
            nil,
            game.get_player(e.player_index)
        )
    end
)
script.on_init(
    function()
        for ____, luaPlayer in pairs(game.players) do
            greetPlayer(nil, luaPlayer)
        end
    end
)"
`;

exports[`helloWorld.ts factorio version 1.1.37 1`] = `
"--[[ Generated with https://github.com/TypeScriptToLua/TypeScriptToLua ]]
function greetPlayer(self, player)
    player.print(\\"Hello, world!\\")
end
script.on_event(
    defines.events.on_player_created,
    function(e)
        greetPlayer(
            nil,
            game.get_player(e.player_index)
        )
    end
)
script.on_init(
    function()
        for ____, luaPlayer in pairs(game.players) do
            greetPlayer(nil, luaPlayer)
        end
    end
)"
`;
